<a href="https://colab.research.google.com/github/monsegutier/Grupo4/blob/main/Anexo_NLP_Churn.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Análisis de opiniones de Clientes**

"He estado con esta compañía durante años, pero últimamente el servicio al cliente ha empeorado mucho. Cada vez que tengo un problema, tardo horas en recibir ayuda."

"Los precios de los planes han subido demasiado sin ninguna mejora notable en los servicios. Estoy considerando cambiarme a otra compañía que ofrezca mejores tarifas."

"La calidad de la señal es inconsistente. En algunas áreas tengo una excelente cobertura, pero en otras es casi inexistente, lo cual es muy frustrante."

"La velocidad de Internet es muy lenta, especialmente durante las horas pico. No puedo realizar mis actividades diarias sin interrupciones constantes."

"Recientemente, he notado cargos adicionales en mi factura que no reconocí. El proceso para disputarlos es complicado y toma mucho tiempo, lo cual me hace sentir descontento con el servicio."

"Estoy muy satisfecho con el servicio que recibo. La calidad de la señal es excelente en todas partes y rara vez tengo interrupciones. Además, las tarifas son competitivas."

"El servicio al cliente es increíble. Cada vez que tengo un problema, el equipo de soporte me ayuda rápidamente y siempre son muy amables. Me siento valorado como cliente.

In [ ]:
#Instalación de librerías
!pip install transformers torch


  Using cached nvidia_cuda_nvrtc_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (23.7 MB)
  Using cached nvidia_cuda_runtime_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (823 kB)
  Using cached nvidia_cuda_cupti_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (14.1 MB)
  Using cached nvidia_cudnn_cu12-8.9.2.26-py3-none-manylinux1_x86_64.whl (731.7 MB)
  Using cached nvidia_cublas_cu12-12.1.3.1-py3-none-manylinux1_x86_64.whl (410.6 MB)
  Using cached nvidia_cufft_cu12-11.0.2.54-py3-none-manylinux1_x86_64.whl (121.6 MB)
  Using cached nvidia_curand_cu12-10.3.2.106-py3-none-manylinux1_x86_64.whl (56.5 MB)
  Using cached nvidia_cusolver_cu12-11.4.5.107-py3-none-manylinux1_x86_64.whl (124.2 MB)
  Using cached nvidia_cusparse_cu12-12.1.0.106-py3-none-manylinux1_x86_64.whl (196.0 MB)
  Using cached nvidia_nccl_cu12-2.20.5-py3-none-manylinux2014_x86_64.whl (176.2 MB)
  Using cached nvidia_nvtx_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (99 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.3/21

In [ ]:
from transformers import pipeline

In [ ]:
# Inicializar el pipeline de análisis de sentimientos
sentiment_pipeline = pipeline("sentiment-analysis")

# Opiniones de clientes
opinions = [
    "He estado con esta compañía durante años, pero últimamente el servicio al cliente ha empeorado mucho. Cada vez que tengo un problema, tardo horas en recibir ayuda.",
    "Los precios de los planes han subido demasiado sin ninguna mejora notable en los servicios. Estoy considerando cambiarme a otra compañía que ofrezca mejores tarifas.",
    "La calidad de la señal es inconsistente. En algunas áreas tengo una excelente cobertura, pero en otras es casi inexistente, lo cual es muy frustrante.",
    "La velocidad de Internet es muy lenta, especialmente durante las horas pico. No puedo realizar mis actividades diarias sin interrupciones constantes.",
    "Recientemente, he notado cargos adicionales en mi factura que no reconocí. El proceso para disputarlos es complicado y toma mucho tiempo, lo cual me hace sentir descontento con el servicio.",
    "Estoy muy satisfecho con el servicio que recibo. La calidad de la señal es excelente en todas partes y rara vez tengo interrupciones. Además, las tarifas son competitivas.",
    "El servicio al cliente es increíble. Cada vez que tengo un problema, el equipo de soporte me ayuda rápidamente y siempre son muy amables. Me siento valorado como cliente."
]

# Analizar el sentimiento de cada opinión
results = sentiment_pipeline(opinions)

# Mostrar los resultados
for opinion, result in zip(opinions, results):
    print(f"Opinion: {opinion}")
    print(f"Sentiment: {result['label']}, Score: {result['score']:.4f}\n")


No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision af0f99b (https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.
/usr/local/lib/python3.10/dist-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that aut

config.json:   0%|          | 0.00/629 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

Opinion: He estado con esta compañía durante años, pero últimamente el servicio al cliente ha empeorado mucho. Cada vez que tengo un problema, tardo horas en recibir ayuda.
Sentiment: NEGATIVE, Score: 0.9130

Opinion: Los precios de los planes han subido demasiado sin ninguna mejora notable en los servicios. Estoy considerando cambiarme a otra compañía que ofrezca mejores tarifas.
Sentiment: NEGATIVE, Score: 0.9311

Opinion: La calidad de la señal es inconsistente. En algunas áreas tengo una excelente cobertura, pero en otras es casi inexistente, lo cual es muy frustrante.
Sentiment: NEGATIVE, Score: 0.9521

Opinion: La velocidad de Internet es muy lenta, especialmente durante las horas pico. No puedo realizar mis actividades diarias sin interrupciones constantes.
Sentiment: NEGATIVE, Score: 0.9666

Opinion: Recientemente, he notado cargos adicionales en mi factura que no reconocí. El proceso para disputarlos es complicado y toma mucho tiempo, lo cual me hace sentir descontento con el 

**Explicación del código**

**Importación de la biblioteca:** Importamos pipeline de la biblioteca transformers.

**Inicialización del pipeline de análisis de sentimientos:** Creamos un pipeline de análisis de sentimientos usando un modelo preentrenado.

**Opiniones de clientes:** Definimos una lista de opiniones  de clientes.

**Análisis de sentimientos:** Usamos el pipeline para analizar cada opinión y obtener el sentimiento (positivo o negativo) y su puntuación de confianza.

**Mostrar resultados:** Imprimimos la opinión original junto con el sentimiento detectado y la puntuación de confianza.

In [ ]:
#Probamos otro modelo, porque las opiniones postivas las toma negativas
# Inicializar el pipeline de análisis de sentimientos en español
sentiment_pipeline = pipeline("sentiment-analysis", model="nlptown/bert-base-multilingual-uncased-sentiment")

# Opiniones de clientes
opinions = [
    "He estado con esta compañía durante años, pero últimamente el servicio al cliente ha empeorado mucho. Cada vez que tengo un problema, tardo horas en recibir ayuda.",
    "Los precios de los planes han subido demasiado sin ninguna mejora notable en los servicios. Estoy considerando cambiarme a otra compañía que ofrezca mejores tarifas.",
    "La calidad de la señal es inconsistente. En algunas áreas tengo una excelente cobertura, pero en otras es casi inexistente, lo cual es muy frustrante.",
    "La velocidad de Internet es muy lenta, especialmente durante las horas pico. No puedo realizar mis actividades diarias sin interrupciones constantes.",
    "Recientemente, he notado cargos adicionales en mi factura que no reconocí. El proceso para disputarlos es complicado y toma mucho tiempo, lo cual me hace sentir descontento con el servicio.",
    "Estoy muy satisfecho con el servicio que recibo. La calidad de la señal es excelente en todas partes y rara vez tengo interrupciones. Además, las tarifas son competitivas.",
    "El servicio al cliente es increíble. Cada vez que tengo un problema, el equipo de soporte me ayuda rápidamente y siempre son muy amables. Me siento valorado como cliente."
]

# Analizar el sentimiento de cada opinión
results = sentiment_pipeline(opinions)

# Mostrar los resultados
for opinion, result in zip(opinions, results):
    print(f"Opinion: {opinion}")
    print(f"Sentiment: {result['label']}, Score: {result['score']:.4f}\n")


config.json:   0%|          | 0.00/953 [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/669M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/39.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/872k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

Opinion: He estado con esta compañía durante años, pero últimamente el servicio al cliente ha empeorado mucho. Cada vez que tengo un problema, tardo horas en recibir ayuda.
Sentiment: 3 stars, Score: 0.3130

Opinion: Los precios de los planes han subido demasiado sin ninguna mejora notable en los servicios. Estoy considerando cambiarme a otra compañía que ofrezca mejores tarifas.
Sentiment: 2 stars, Score: 0.5965

Opinion: La calidad de la señal es inconsistente. En algunas áreas tengo una excelente cobertura, pero en otras es casi inexistente, lo cual es muy frustrante.
Sentiment: 2 stars, Score: 0.5137

Opinion: La velocidad de Internet es muy lenta, especialmente durante las horas pico. No puedo realizar mis actividades diarias sin interrupciones constantes.
Sentiment: 2 stars, Score: 0.5696

Opinion: Recientemente, he notado cargos adicionales en mi factura que no reconocí. El proceso para disputarlos es complicado y toma mucho tiempo, lo cual me hace sentir descontento con el serv

El análisis de sentimientos realizado con el modelo nlptown/bert-base-multilingual-uncased-sentiment genera una clasificación de estrellas (de 1 a 5) junto con una puntuación de confianza para cada opinión. Aquí está la interpretación de cada resultado:

**Opinión: "He estado con esta compañía durante años, pero últimamente el servicio al cliente ha empeorado mucho. Cada vez que tengo un problema, tardo horas en recibir ayuda.**"

Sentiment: 3 stars
Score: 0.3130

**Interpretación:** Aunque la opinión menciona problemas recientes con el servicio al cliente, la clasificación de 3 estrellas indica una evaluación moderada o neutral. La puntuación relativamente baja sugiere una confianza moderada en esta clasificación.

**Opinión: "Los precios de los planes han subido demasiado sin ninguna mejora notable en los servicios. Estoy considerando cambiarme a otra compañía que ofrezca mejores tarifas."**

Sentiment: 2 stars
Score: 0.5965

**Interpretación:** La clasificación de 2 estrellas refleja una evaluación negativa debido a los aumentos de precios sin mejoras en los servicios. La puntuación de confianza es relativamente alta, indicando una confianza considerable en esta evaluación negativa.

**Opinión: "La calidad de la señal es inconsistente. En algunas áreas tengo una excelente cobertura, pero en otras es casi inexistente, lo cual es muy frustrante."**

Sentiment: 2 stars
Score: 0.5137

**Interpretación:** La inconsistencia en la calidad de la señal lleva a una clasificación negativa de 2 estrellas. La puntuación de confianza es moderada, sugiriendo una confianza razonable en la evaluación negativa.

**Opinión: "La velocidad de Internet es muy lenta, especialmente durante las horas pico. No puedo realizar mis actividades diarias sin interrupciones constantes."**

Sentiment: 2 stars
Score: 0.5696

**Interpretación:** La frustración con la velocidad de Internet resulta en una clasificación de 2 estrellas. La puntuación de confianza es relativamente alta, indicando una evaluación negativa confiable.

**Opinión: "Recientemente, he notado cargos adicionales en mi factura que no reconocí. El proceso para disputarlos es complicado y toma mucho tiempo, lo cual me hace sentir descontento con el servicio."**

Sentiment: 2 stars
Score: 0.5341

**Interpretación**: Los cargos no reconocidos y el complicado proceso de disputa llevan a una evaluación negativa de 2 estrellas. La puntuación de confianza es moderada, sugiriendo una evaluación negativa razonable.

**Opinión: "Estoy muy satisfecho con el servicio que recibo. La calidad de la señal es excelente en todas partes y rara vez tengo interrupciones. Además, las tarifas son competitivas."**

Sentiment: 5 stars
Score: 0.6876

**Interpretación: **La satisfacción con la calidad de la señal y las tarifas competitivas resulta en una evaluación muy positiva de 5 estrellas. La puntuación de confianza es alta, indicando una evaluación positiva confiable.

**Opinión: "El servicio al cliente es increíble. Cada vez que tengo un problema, el equipo de soporte me ayuda rápidamente y siempre son muy amables. Me siento valorado como cliente."**

Sentiment: 5 stars
Score: 0.8130

**Interpretación:** La excelente experiencia con el servicio al cliente lleva a una evaluación muy positiva de 5 estrellas. La puntuación de confianza es muy alta, indicando una evaluación extremadamente positiva y confiable.
Las opiniones negativas tienen una puntuación de 2 a 3 estrellas con una confianza moderada a alta en la evaluación, mientras que las opiniones positivas tienen una puntuación de 5 estrellas con una alta confianza en la evaluación. Esto permite identificar claramente las áreas de insatisfacción y satisfacción entre los clientes.







In [ ]:
from transformers import MarianMTModel, MarianTokenizer


Vamos a traducir las opiniones de los clientes a inglés

In [ ]:
# Función para traducir usando el modelo de MarianMT
def translate_to_english(texts):
    model_name = 'Helsinki-NLP/opus-mt-es-en'
    tokenizer = MarianTokenizer.from_pretrained(model_name)
    model = MarianMTModel.from_pretrained(model_name)

    translated = []
    for text in texts:
        translated_text = model.generate(**tokenizer(text, return_tensors="pt", padding=True))
        translated.append(tokenizer.decode(translated_text[0], skip_special_tokens=True))
    return translated

# Opiniones de clientes (en español)
opinions = [
    "He estado con esta compañía durante años, pero últimamente el servicio al cliente ha empeorado mucho. Cada vez que tengo un problema, tardo horas en recibir ayuda.",
    "Los precios de los planes han subido demasiado sin ninguna mejora notable en los servicios. Estoy considerando cambiarme a otra compañía que ofrezca mejores tarifas.",
    "La calidad de la señal es inconsistente. En algunas áreas tengo una excelente cobertura, pero en otras es casi inexistente, lo cual es muy frustrante.",
    "La velocidad de Internet es muy lenta, especialmente durante las horas pico. No puedo realizar mis actividades diarias sin interrupciones constantes.",
    "Recientemente, he notado cargos adicionales en mi factura que no reconocí. El proceso para disputarlos es complicado y toma mucho tiempo, lo cual me hace sentir descontento con el servicio.",
    "Estoy muy satisfecho con el servicio que recibo. La calidad de la señal es excelente en todas partes y rara vez tengo interrupciones. Además, las tarifas son competitivas.",
    "El servicio al cliente es increíble. Cada vez que tengo un problema, el equipo de soporte me ayuda rápidamente y siempre son muy amables. Me siento valorado como cliente."
]

# Traducir al inglés (es -> en)
translated_to_english = translate_to_english(opinions)
print("Opiniones en inglés:")
for opinion in translated_to_english:
    print(opinion)


Opiniones en inglés:
I've been with this company for years, but lately customer service has gotten a lot worse every time I have a problem, it takes me hours to get help.
Plan prices have risen too much without any noticeable improvement in services. I am considering switching to another company that offers better rates.
The quality of the signal is inconsistent. In some areas I have excellent coverage, but in others it is almost non-existent, which is very frustrating.
Internet speed is very slow, especially during peak hours. I cannot perform my daily activities without constant interruptions.
Recently, I have noticed additional charges on my bill that I did not recognize. The process of disputing them is complicated and takes a long time, which makes me feel unhappy with the service.
I am very satisfied with the service I receive. The quality of the signal is excellent everywhere and I rarely have interruptions. In addition, the rates are competitive.
The customer service is amazing

Traducción a francés

In [ ]:
# Función para traducir usando el modelo de MarianMT
def translate_to_french(texts):
    model_name = 'Helsinki-NLP/opus-mt-es-fr'
    tokenizer = MarianTokenizer.from_pretrained(model_name)
    model = MarianMTModel.from_pretrained(model_name)

    translated = []
    for text in texts:
        translated_text = model.generate(**tokenizer(text, return_tensors="pt", padding=True))
        translated.append(tokenizer.decode(translated_text[0], skip_special_tokens=True))
    return translated

# Opiniones de clientes (en español)
opinions = [
    "He estado con esta compañía durante años, pero últimamente el servicio al cliente ha empeorado mucho. Cada vez que tengo un problema, tardo horas en recibir ayuda.",
    "Los precios de los planes han subido demasiado sin ninguna mejora notable en los servicios. Estoy considerando cambiarme a otra compañía que ofrezca mejores tarifas.",
    "La calidad de la señal es inconsistente. En algunas áreas tengo una excelente cobertura, pero en otras es casi inexistente, lo cual es muy frustrante.",
    "La velocidad de Internet es muy lenta, especialmente durante las horas pico. No puedo realizar mis actividades diarias sin interrupciones constantes.",
    "Recientemente, he notado cargos adicionales en mi factura que no reconocí. El proceso para disputarlos es complicado y toma mucho tiempo, lo cual me hace sentir descontento con el servicio.",
    "Estoy muy satisfecho con el servicio que recibo. La calidad de la señal es excelente en todas partes y rara vez tengo interrupciones. Además, las tarifas son competitivas.",
    "El servicio al cliente es increíble. Cada vez que tengo un problema, el equipo de soporte me ayuda rápidamente y siempre son muy amables. Me siento valorado como cliente."
]

# Traducir al francés (es -> fr)
translated_to_french = translate_to_french(opinions)
print("Opiniones en francés:")
for opinion in translated_to_french:
    print(opinion)


tokenizer_config.json:   0%|          | 0.00/42.0 [00:00<?, ?B/s]

source.spm:   0%|          | 0.00/819k [00:00<?, ?B/s]

target.spm:   0%|          | 0.00/812k [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/2.34M [00:00<?, ?B/s]

config.json:   0%|          | 0.00/1.38k [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/332M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/293 [00:00<?, ?B/s]

Opiniones en francés:
J'ai été avec cette société pendant des années, mais ces derniers temps le service à la clientèle a empiré beaucoup. Chaque fois que j'ai un problème, il me faut des heures pour recevoir de l'aide.
Les prix des plans ont trop augmenté sans aucune amélioration notable dans les services. J'envisage de passer à une autre entreprise offrant de meilleurs tarifs.
La qualité du signal est incohérente. Dans certains domaines, j'ai une excellente couverture, mais dans d'autres, elle est presque inexistante, ce qui est très frustrant.
La vitesse d'Internet est très lente, surtout pendant les heures de pointe. Je ne peux pas faire mes activités quotidiennes sans interruption constante.
Récemment, j'ai remarqué des frais supplémentaires sur ma facture que je n'ai pas reconnu. Le processus pour les disputer est compliqué et prend beaucoup de temps, ce qui me rend insatisfait avec le service.
Je suis très satisfait du service que je reçois. La qualité du signal est excellente p